### PR TITLE
[Merged by Bors] - feat(FieldTheory/SeparableDegree): `Field.Emb` is infinite for transcendental extension

### PR DESCRIPTION
**Pull request by acmepjz**
- `Field.infinite_emb_of_transcendental`, `Field.finSepDegree_eq_zero_of_transcendental`: `Field.Emb` is infinite and `Field.finSepDegree` is zero for transcendental extensions
- remove algebraic assumptions in `isPurelyInseparable_(of|iff)_finSepDegree_eq_one`
- prove `IsPurelyInseparable.of_injective_comp_algebraMap`
- change `Field.Emb` from `def` to `abbrev`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

- [x] depends on: #18648
- [x] depends on: #18769

### CODE REVIEW

File: Mathlib/FieldTheory/SeparableDegree.lean [hunk 1]
```diff
@@ -246,6 +249,44 @@ def embProdEmbOfIsAlgebraic [Algebra E K] [IsScalarTower F E K] [Algebra.IsAlgeb
...
+            (Polynomial.transcendental_X F).pow (Nat.succ_pos n)
+  have hf : Function.Injective f := fun m n h ↦ by
+    replace h : (MvPolynomial.X i) ^ (m + 1) = (MvPolynomial.X i) ^ (n + 1) := hi1 <| by
+      simpa [f, -map_pow] using congr($h (algebraMap _ (FractionRing (MvPolynomial ι F))
+        (MvPolynomial.X (R := F) i)))
+    apply_fun MvPolynomial.coeff (Finsupp.single i (m + 1)) at h
+    classical simp_rw [MvPolynomial.coeff_single_X_pow] at h
+    by_contra!
+    simp [if_neg this.symm] at h
+  exact Infinite.of_injective f hf
```
Comment (alreadydone) [inline Mathlib/FieldTheory/SeparableDegree.lean:283]:
```suggestion
  have := hx.isAlgebraic_field
  rw [← (embProdEmbOfIsAlgebraic F (adjoin F (Set.range x)) E).infinite_iff]
  refine @Prod.infinite_of_left _ _ ?_ _
  rw [← (embEquivOfEquiv _ _ _ hx.1.aevalEquivField).infinite_iff]
  obtain ⟨i⟩ := hx.nonempty_iff_transcendental.2 H
  let K := FractionRing (MvPolynomial ι F)
  let i1 := IsScalarTower.toAlgHom F (MvPolynomial ι F) (AlgebraicClosure K)
  have hi1 : Function.Injective i1 := by
    rw [IsScalarTower.coe_toAlgHom', IsScalarTower.algebraMap_eq _ K]
    exact (algebraMap K (AlgebraicClosure K)).injective.comp (IsFractionRing.injective _ _)
  let f (n : ℕ) : Emb F K := IsFractionRing.liftAlgHom
    (g := i1.comp <| MvPolynomial.aeval fun i : ι ↦ MvPolynomial.X i ^ (n + 1)) <| hi1.comp <| by
      simpa [algebraicIndependent_iff_injective_aeval] using
        MvPolynomial.algebraicIndependent_polynomial_aeval_X _
          fun i : ι ↦ (Polynomial.transcendental_X F).pow n.succ_pos
  refine Infinite.of_injective f fun m n h ↦ ?_
  replace h : (MvPolynomial.X i) ^ (m + 1) = (MvPolynomial.X i) ^ (n + 1) := hi1 <| by
    simpa [f, -map_pow] using congr($h (algebraMap _ K (MvPolynomial.X (R := F) i)))
  simpa using congr(MvPolynomial.totalDegree $h)
```

File: Mathlib/FieldTheory/PurelyInseparable.lean [hunk 2]
```diff
@@ -683,15 +680,25 @@ theorem eq_separableClosure_iff [Algebra.IsAlgebraic F E] (L : IntermediateField
...
+  rw [Function.Injective] at h
+  contrapose! h
+  obtain ⟨f, g, hfg⟩ := exists_pair_ne (Emb F E)
+  obtain ⟨i⟩ := hn
+  letI := i.toAlgebra
+  let j : AlgebraicClosure E →ₐ[E] L := IsAlgClosed.lift
+  refine ⟨j.toRingHom.comp f.toRingHom, j.toRingHom.comp g.toRingHom, by ext; simp, fun H ↦ ?_⟩
+  apply_fun (fun x ↦ (x : E → L)) at H
+  simp_rw [AlgHom.toRingHom_eq_coe, RingHom.coe_comp, RingHom.coe_coe] at H
+  exact hfg (AlgHom.coe_fn_inj.1 (j.injective.comp_left H))
```
Comment (alreadydone) [inline Mathlib/FieldTheory/PurelyInseparable.lean:701]:
```suggestion
    [Nonempty (E →+* L)] (h : Function.Injective fun f : E →+* L ↦ f.comp (algebraMap F E)) :
    IsPurelyInseparable F E := by
  rw [isPurelyInseparable_iff_finSepDegree_eq_one, finSepDegree, Nat.card_eq_one_iff_unique]
  letI := (Classical.arbitrary (E →+* L)).toAlgebra
  let j : AlgebraicClosure E →ₐ[E] L := IsAlgClosed.lift
  exact ⟨⟨fun f g ↦ DFunLike.ext' <| j.injective.comp_left (congr_arg (⇑) <|
    @h (j.toRingHom.comp f) (j.toRingHom.comp g) (by ext; simp))⟩, inferInstance⟩
```

**Review by alreadydone — APPROVED** [review]
Thanks, LGTM!
It turns out #9392 is not needed, and [Field.embProdEmbOfIsAlgebraic](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/SeparableDegree.html#Field.embProdEmbOfIsAlgebraic) is enough.
Here are some golfs:

**Comment by alreadydone** [comment]
maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**